### ISSUE TITLE
odd NoneType behaviour

### ISSUE THREAD

**Issue by jrmlhermitte** [bluesky/databroker#302]
I can't figure out what is wrong, nor why this is occurring for now, but I'll document as best as I can for now. Seems to happen only for the specific case I show here.
```python
In [1]: hdrs = cmsdb(start_time="2017-06-01", stop_time="2017-06-02")

In [2]: stream = iter(cmsdb.restream(hdrs))

In [3]: next(stream)

```
gives:
```python
TypeError: 'NoneType' object is not iterable
```

However, with @danielballan 's new PR, it works just fine (PR #300):
```python
In [1]: hdrs = cmsdb(start_time="2017-06-01", stop_time="2017-06-02")

In [2]: stream = iter(cmsdb.restream(hdrs))

In [3]: next(stream)
Out[3]: 
('start',
 {'beam_divergence_x_mrad': 0.09994800000000001,
  'beam_divergence_y_mrad': 0.09999100000000001,
  'beam_intensity_expected': 11000,
[...]

In [4]: next(stream)
Out[4]: 
('stop',
 {'exit_status': 'success',
  'run_start': '93cc8bf6-b0a9-46ca-9e12-fcac153214ba',
  'time': 1496338481.141322,
  'uid': '38397b8d-7b75-444f-a809-f4f5d5a48629'})

In [5]: next(stream)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call las
[...]
```

It seems to me that @danielballan 's fix may be skipping this one element? I am a little confused, and this may be an error on my part. But I don't want to lose track of this as it could point to something bigger.
When I have time, I'll try looking further. If anyone has insight as of now though, it would be appreciated.

thanks!

**Comment by tacaswell** [comment]
Can you drop into `%debug` and walk up the tree to sort where we are trying to iterate on a `None`?

I suspect that there is a run without a `Stop` document and we have waffled on if a missing stop document should be `None` or `{}` by default.

**Comment by jrmlhermitte** [comment]
yep, i think you're right:
```python
In [4]: %debug
> /home/lhermitte/anaconda3/envs/py3/lib/python3.6/site-packages/doct-1.0.0-py3.6.egg/doct.py(73)__init__()
     71 
     72     def __init__(self, name, *args, **kwargs):
---> 73         super(Document, self).__init__(*args, **kwargs)
     74         super(Document, self).__setitem__('_name', name)
     75         super(Document, self).__setattr__('__dict__', self)

ipdb> u
> /home/lhermitte/research/projects/databroker/databroker/_core.py(1051)wrap_in_deprecated_doct()
   1049     in the future.
   1050     """
-> 1051     return DeprecatedDoct(DOCT_NAMES[name], doc)
   1052 
   1053 

ipdb> p name
'stop'
ipdb> p doc
None
ipdb> p DeprecatedDoct('stop', {})
{'_name': 'stop'}
ipdb> p DeprecatedDoct('stop', None)
*** TypeError: 'NoneType' object is not iterable
ipdb> 

```
so should it be `{}` by default as you said?

**Comment by jrmlhermitte** [comment]
oh and the full backtrace:
```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-2799681e0bef> in <module>()
----> 1 next(stream)

~/research/projects/databroker/databroker/_core.py in restream(self, headers, fields, fill)
   1754         :meth:`Broker.process`
   1755         """
-> 1756         for payload in self.get_documents(headers, fields=fields, fill=fill):
   1757             yield payload
   1758 

~/research/projects/databroker/databroker/_core.py in get_documents(self, headers, stream_name, fields, fill, handler_registry)
   1529             headers = [headers]
   1530 
-> 1531         check_fields_exist(fields if fields else [], headers)
   1532         # dirty hack!
   1533         with self.reg.handler_context(handler_registry):

~/research/projects/databroker/databroker/eventsource/shim.py in check_fields_exist(fields, headers)
    444     comp_re = _compile_re(fields)
    445     all_fields = set()
--> 446     for header in headers:
    447         all_fields.update(header['start'])
    448         all_fields.update(header.get('stop', {}))
~/research/projects/databroker/databroker/_core.py in __iter__(self)
    844         for start, stop in res:
    845             header = Header(start=self._db.prepare_hook('start', start),
--> 846                             stop=self._db.prepare_hook('stop', stop),
    847                             db=self._db)
    848             if self._data_key is None:

~/research/projects/databroker/databroker/_core.py in wrap_in_deprecated_doct(name, doc)
   1049     in the future.
   1050     """
-> 1051     return DeprecatedDoct(DOCT_NAMES[name], doc)
   1052 
   1053 

~/anaconda3/envs/py3/lib/python3.6/site-packages/doct-1.0.0-py3.6.egg/doct.py in __init__(self, name, *args, **kwargs)
     71 
     72     def __init__(self, name, *args, **kwargs):
---> 73         super(Document, self).__init__(*args, **kwargs)
     74         super(Document, self).__setitem__('_name', name)
     75         super(Document, self).__setattr__('__dict__', self)

TypeError: 'NoneType' object is not iterable

```

**Comment by tacaswell** [comment]
A quick fix is to change line 846 to `stop=self._db.prepare_hook('stop', stop or {}),` but it maybe should be fixed deeper down in HeaderSource

**Comment by jrmlhermitte** [comment]
yeah, and a test should definitely be added. I could start a PR a little later if it helps. Thanks for the support!

**Comment by jrmlhermitte** [comment]
discussion moved to #303

**Comment by danielballan** [comment]
Closed by #303.
